lib/ASTGen/Sources/ASTGen/LexicalLookup.swift

```diff
@@ -0,0 +1,611 @@
+
+  return astScopeResultArray.compactMap { bridgedResult in
+    let identifierPointer = bridgedResult.name.raw?.assumingMemoryBound(to: CChar.self)
+
+    guard let astResultPosition = sourceFile.pointee.position(of: bridgedResult.nameLoc) else {
+      print("One of the results, doesn't have a position")
+      return nil
+    }
+
+    let consumedResult = ConsumedLookupResult(
+      rawName: identifierPointer == nil ? "" : String(cString: identifierPointer!),
+      position: astResultPosition,
+      flags: ConsumedLookupResultFlag(rawValue: bridgedResult.flag)
+    )
+
+    // If the name doesn't have any flags and
+    // the name is empty, it should be omitted.
+    if consumedResult.flags.isEmpty && consumedResult.name.isEmpty {
+      consumedResult.flags.insert(.shouldBeOmitted)
+    }
+
+    return consumedResult
+  }
+}
+
+/// Performs and returns `SwiftLexicalLookup` lookup and returns
+/// the results an array of `ConsumedLookupResult`. Introduces appropriate flags.
+private func sllConsumedResults(
+  lookupToken: TokenSyntax,
+  finishInSequentialScope: Bool,
+  buildConfiguration: CompilerBuildConfiguration
+) -> [ConsumedLookupResult] {
+  lookupToken.lookup(
+    nil,
+    with: LookupConfig(finishInSequentialScope: finishInSequentialScope, buildConfiguration: buildConfiguration)
+  )
+  .flatMap { result in
+    switch result {
+    case .lookInMembers(let lookInMembers):
+      return [
+        ConsumedLookupResult(
+          rawName: "",
+          position: lookInMembers.lookupMembersPosition,
+          flags: .shouldLookInMembers
+        )
+      ]
+    case .lookInGenericParametersOfExtendedType(let extensionDecl):
+      return [
+        ConsumedLookupResult(
+          rawName: "",
+          position: extensionDecl.extensionKeyword.positionAfterSkippingLeadingTrivia,
+          flags: .ignoreNextFromHere
+        )
+      ]
+    case .mightIntroduceDollarIdentifiers(let closure):
+      return [
+        ConsumedLookupResult(
+          rawName: "",
+          position: closure.positionAfterSkippingLeadingTrivia,
+          flags: .ignoreNextFromHere
+        )
+      ]
+    default:
+      if let parent = result.scope.parent, result.scope.is(GenericParameterClauseSyntax.self) {
+        if let parentFunctionDecl = parent.as(FunctionDeclSyntax.self),
+          parentFunctionDecl.attributes.range.contains(lookupToken.position)
+        {
+          // If lookup started from inside function attributes, don't reverse.
+          return result.names.map { name in
+            ConsumedLookupResult(rawName: name.identifier?.name ?? "", position: name.position, flags: [])
+          }
+        } else if parent.is(FunctionDeclSyntax.self) || parent.is(SubscriptDeclSyntax.self)
+          || result.scope.range.contains(lookupToken.position)
+        {
+          // If a result from function generic parameter clause or lookup started within it, reverse introduced names.
```

Comment:
I guess the old code has some funky behaviors :)

Reply:
Yes, we've encountered some really fun behavior of `ASTScope` when working on the library.

This particular case happens when looking up generic parameter from generic parameter clause which is part of a function or subscript or the lookup is started within the clause:
```swift
func foo<A, B: Int>(a: A, b: B) {
  print(a, b)
}

```
`ASTScope` for some reason decides to first introduce generic parameters in wrong order just to introduce them correctly right afterwards :). Interestingly, similar behavior doesn't happen inside nominal type declarations.
```
-----> Lookup started at: 2:3 ("print") finishInSequentialScope: false
     |           ASTScope           |      SwiftLexicalLookup      
> ✅ |            a 1:21            |            a 1:21            
> ✅ |            b 1:27            |            b 1:27            
> ✅ |            B 1:13            |           ↕️ B 1:13           
> ✅ |            A 1:10            |           ↕️ A 1:10           
> ℹ️ | Omitted ASTScope name: A 1:10
> ℹ️ | Omitted ASTScope name: B 1:13
```

With those comments, this validation function slowly becomes a documentation of "weird" `ASTScope` behavior. It might be necessary to revisit some of those heuristics in the future if it turns out there's a valid reason for this behavior and we should actually adjust `SwiftLexicalLookup`. That's why I tried to be more explicit in this part of the validation algorithm.